lib/util/extractSourceMap.js

```diff
@@ -0,0 +1,290 @@
+			return fetchFromDataURL(url, (err, sourceContent) => {
+				if (err) return callback(err);
+				callback(null, { sourceURL: "", sourceContent });
+			});
+		} else if (protocol === "file:") {
+			const sourceURL = urlUtils.fileURLToPath(url);
+			if (skipReading) return callback(null, { sourceURL });
+			return fetchFromFilesystem(fs, sourceURL, (err, result) => {
+				if (err) return callback(err);
+				const { data: sourceContent } = result;
+				callback(null, { sourceURL, sourceContent });
+			});
+		} else {
+			return callback(
+				new Error(`Failed to parse source map: '${url}' URL is not supported`)
+			);
+		}
+	}
+
+	// 2. It's a scheme-relative
+	const doubleSlashStartPattern = /^\/\//;
+	if (doubleSlashStartPattern.test(url)) {
+		return callback(
+			new Error(`Failed to parse source map: '${url}' URL is not supported`)
+		);
+	}
+
+	// 3. Absolute path
+	if (isAbsolute(url)) {
+		if (skipReading) return callback(null, { sourceURL: url });
+		const possibleRequests = [url];
+
+		if (url.startsWith("/") || url.startsWith("\\")) {
+			possibleRequests.push(getAbsolutePath(fs, context, url.slice(1)));
+		}
+
+		fetchPathsFromFilesystem(fs, possibleRequests, "", (err, result) => {
+			if (err) return callback(err);
+			const { path: sourceURL, data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	} else {
+		// 4. Relative path
+		const sourceURL = getAbsolutePath(fs, context, url);
+		if (skipReading) return callback(null, { sourceURL });
+
+		fetchFromFilesystem(fs, sourceURL, (err, result) => {
+			if (err) return callback(err);
+			const { data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	}
+}
+
+/**
+ * @param {string} input input data
+ * @param {InputFileSystem} fs filesystem
+ * @param {NormalModule} module normal module
+ * @param {boolean} removeOnly should only remove source map comment
+ * @param {(err: Error|null, code?: string, sourceMap?: object) => void} callback callback
+ */
+module.exports = function extractSourceMap(
+	input,
+	fs,
+	module,
+	removeOnly,
+	callback
+) {
+	const { sourceMappingURL, replacementString } = getSourceMappingURL(input);
+
+	if (!sourceMappingURL) {
+		callback(null, input);
+		return;
+	} else if (removeOnly) {
+		callback(null, input.replace(replacementString, ""));
+		return;
+	}
+
+	fetchFromURL(
+		fs,
+		module.context || "",
+		sourceMappingURL,
+		false,
+		(err, result) => {
+			if (err) return callback(err);
+			const { sourceURL, sourceContent } = result;
+
+			if (sourceURL && module.buildInfo && module.buildInfo.fileDependencies) {
+				module.buildInfo.fileDependencies.add(sourceURL);
+			}
+
+			let map;
+
+			try {
+				map = JSON.parse(sourceContent.replace(/^\)\]\}'/, ""));
+			} catch (parseError) {
+				return callback(
+					new WebpackError(
+						`Failed to parse source map from '${sourceMappingURL}': ${parseError}`
+					)
+				);
```

Comment:
Based on feedback during development of https://github.com/webpack-contrib/source-map-loader/, better to make an options for this, in real world a lot of packages have broken source maps and developer can't fix it in vendor libraries, so better generate a warning, because it really doesn't break a build, but allow to setup it to make an error, i.e.

```ts
type ExtractSourceMap = boolean | { errorSeverity: "none" | "warn" | "error" }
```